Fix global namespace pollution in ATen/Dispatch.h

slyfox3 opened the pull request:
Summary:

Was it a typo? Since we already have `at::detail::record_kernel_function_dtype()` in `ATen/Dispatch.h`

Test Plan: just build

Differential Revision: D64642080

Skylion007 commented on aten/src/ATen/Dispatch.h:
```diff
@@ -98,6 +98,7 @@ TORCH_API void record_kernel_function_dtype(std::string name);
     return __VA_ARGS__();                                                   \
   }
 
+namespace at {
```
nit: use concatenated namespaces. We are using C++17 now after alll

slyfox3 replied:
(won't fix) Thanks for the comment @Skylion007. I changed from concatenated namespaces back to current form to be consistent with the rest of the file.

malfet replied:
This includes a bit involved patch internally, imo we should split it in a few patches, one that move those to `at::detail` namespace but keeps old macros there for BC pruposes...

slyfox3 commented:
`@pytorchbot label "topic: not user facing"`

slyfox3 commented:
@pytorchbot label "topic: not user facing"

slyfox3 commented:
/easycla

slyfox3 commented:
Merge conflict re-introduced by [PR139358t](https://github.com/pytorch/pytorch/pull/139358) :(

malfet commented on aten/src/ATen/Dispatch.h:
```diff
@@ -193,20 +185,19 @@ inline at::ScalarType scalar_type(at::ScalarType s) {
-    const auto& the_type = TYPE;                                            \
+    const at::ScalarType the_type = TYPE;                                   \
     constexpr const char* at_dispatch_name = NAME;                          \
     /* don't use TYPE again in case it is an expensive or side-effect op */ \
```
This comment is out of place here now, why instead of keeping definition of `the_type` not say `at::ScalarType _st = TYPE;` and this will make this patch even smaller

slyfox3 committed:
Fix global namespace pollution in ATen/Dispatch.h (#138626)

Summary:

Was it a typo? Since we already have `at::detail` in `ATen/Dispatch.h`

Deleting `::detail` all together, we don't need the functions in it anymore.

Test Plan: just build

Reviewed By: malfet

Differential Revision: D64642080

malfet commented:
@pytorchbot revert -m "This broke inductor tests, see https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=inductor_torchbench%2C%202%2C%202" -c nosignal

slyfox3 committed:
Merge branch 'pytorch:main' into export-D64642080
